golem-worker-executor-base/src/invocation.rs

```diff
@@ -407,6 +425,118 @@ async fn invoke<Ctx: WorkerCtx>(
     }
 }
 
+async fn invoke_http_handler<Ctx: WorkerCtx>(
+    store: &mut impl AsContextMut<Data = Ctx>,
+    instance: &wasmtime::component::Instance,
+    function_input: &[Value],
+    raw_function_name: &str,
+) -> Result<InvokeResult, GolemError> {
+    let (sender, receiver) = tokio::sync::oneshot::channel();
+
+    let proxy = Proxy::new(&mut *store, instance).unwrap();
+    let mut store_context = store.as_context_mut();
+
+    store_context.data_mut().borrow_fuel().await?;
+
+    let idempotency_key = store_context.data().get_current_idempotency_key().await;
+    if let Some(idempotency_key) = &idempotency_key {
+        store_context
+            .data()
+            .get_public_state()
+            .event_service()
+            .emit_invocation_start(
+                raw_function_name,
+                idempotency_key,
+                store_context.data().is_live(),
+            );
+    }
+
+    let (_, mut task_exits) = {
+        let hyper_request =
+            virtual_export_compat::http_incoming_handler::input_to_hyper_request(function_input)?;
+        let scheme = wasi_http_scheme_from_request(&hyper_request)?;
+        let incoming = store_context
+            .data_mut()
+            .durable_ctx_mut()
+            .as_wasi_http_view()
+            .new_incoming_request(scheme, hyper_request)
+            .unwrap();
+        let outgoing = store_context
+            .data_mut()
+            .durable_ctx_mut()
+            .as_wasi_http_view()
+            .new_response_outparam(sender)
+            .unwrap();
+
+        unsafe {
```

Comment:
I couldn't figure out from reading it the need of unsafe here. Would you mind explaining it here?

Reply:
this comes from the async_scoped crate. https://docs.rs/async-scoped/latest/async_scoped/

The invariant they want here is just that:
```
An asynchronous function that creates a scope and immediately awaits the stream. The outputs of the futures are collected as a Vec and returned along with the output of the block.
Safety

This function is not completely safe: please see cancellation_soundness in [tests.rs](https://github.com/rmanoka/async-scoped/blob/master/src/tests.rs) for a test-case that suggests how this can lead to invalid memory access if not dealt with care.

The caller must ensure that the lifetime ’a is valid until the returned future is fully driven. Dropping the future is okay, but blocks the current thread until all spawned futures complete.
```